src/App/Pages/Accounts/HomePageViewModel.cs

```diff
@@ -165,39 +163,47 @@ await _stateService.IsAuthenticatedAsync(userId))
             }
         }
 
-        public async Task ShowEnvironmentPicker()
+        public async Task ShowEnvironmentPickerAsync()
         {
-            var options = new string[] { AppResources.US, AppResources.SelfHosted };
-            if (DisplayEuEnvironment)
-            {
-                options = new string[] { AppResources.US, AppResources.EU, AppResources.SelfHosted };
-            }
+            var options = _displayEuEnvironment
+                    ? new string[] { AppResources.US, AppResources.EU, AppResources.SelfHosted }
+                    : new string[] { AppResources.US, AppResources.SelfHosted };
 
-            var result = await Page.DisplayActionSheet(AppResources.DataRegion, AppResources.Cancel, null, options);
-            if (result == AppResources.US)
-            {
-                await _environmentService.SetUsUrlsAsync();
-                SelectedEnvironmentName = AppResources.US;
-            }
-            else if (result == AppResources.EU)
-            {
-                await _environmentService.SetEuUrlsAsync();
-                SelectedEnvironmentName = AppResources.EU;
-            }
-            else if (result == AppResources.SelfHosted)
-            {
-                StartEnvironmentAction?.Invoke();
-            }
+            await Device.InvokeOnMainThreadAsync(async () => {
+                var result = await Page.DisplayActionSheet(AppResources.DataRegion, AppResources.Cancel, null, options);
+
+                if (result is null || result == AppResources.Cancel)
+                {
+                    return;
+                }
+
+                if (result == AppResources.SelfHosted)
+                {
+                    StartEnvironmentAction?.Invoke();
+                }
+                else 
+                {
+                    await _environmentService.SetUrlsAsync( result == AppResources.EU ? EnvironmentUrlData.DefaultEU : EnvironmentUrlData.DefaultUS);
+                    SelectedEnvironmentName = result;
+                }
+            });
         }
 
         public async Task UpdateEnvironment()
         {
-            var baseUrlSaved = (await _stateService.GetPreAuthEnvironmentUrlsAsync())?.Base;
-            if (baseUrlSaved == _environmentService.DefaultBaseUsRegion || string.IsNullOrEmpty(baseUrlSaved))
+            var environmentsSaved = await _stateService.GetPreAuthEnvironmentUrlsAsync();
+            if (environmentsSaved == null || environmentsSaved.Equals(new EnvironmentUrlData()))
```

Comment:
⚠️ I see you override `Equals(...)` to compare if the saved one is empty basically. I wouldn't do that for this requirement, I think it's a little too much, and you're allocating a new object just to compare this (also when you override `Equals(...)` you should also override `GetHashCode(...)` or a hard to look for bug can happen if the object is used in lists)
I'd remove the `Equals(...)` and would do it simpler by just adding a readonly property in `EnvironmentUrlData` -> `IsEmpty`:
```c#
public bool IsEmpty => string.IsNullOrWhiteSpace(Base)
    && string.IsNullOrWhiteSpace(Api)
    && string.IsNullOrWhiteSpace(Identity)
    && string.IsNullOrWhiteSpace(Icons)
    && string.IsNullOrWhiteSpace(Notifications)
    && string.IsNullOrWhiteSpace(WebVault)
    && string.IsNullOrWhiteSpace(Events);
```

❓ Which would be the case the `EnvironmentUrlData` is not `null` but empty? Is that when the user filled the env data and then edited it and clear it out?